app/src/hooks/useCompoundService.tsx

```diff
@@ -0,0 +1,45 @@
+import { useEffect, useMemo, useState } from 'react'
+
+import { CompoundService } from '../services'
+import { CompoundTokenType, Token } from '../util/types'
+
+import { ConnectedWeb3Context } from './connectedWeb3'
+
+export const useCompoundService = (
+  collateral: Token,
+  context: ConnectedWeb3Context,
+): {
+  compoundService: Maybe<CompoundService>
+  fetchCompoundService: () => Promise<void>
+} => {
+  const { account, library: provider } = context
+
+  const [compoundService, setCompoundService] = useState<Maybe<CompoundService>>(null)
+
+  const fetchCompoundService = async () => {
+    if (compoundService) {
+      await compoundService.init()
+      setCompoundService(compoundService)
+    }
+  }
+
+  useMemo(() => {
+    if (collateral.symbol.toLowerCase() in CompoundTokenType) {
+      const compoundService = new CompoundService(collateral.address, collateral.symbol, provider, account)
+      setCompoundService(compoundService)
+      fetchCompoundService()
+    }
+    // eslint-disable-next-line
+  }, [collateral.address, account, collateral.symbol, provider])
+
+  useEffect(() => {
+    if (collateral.symbol.toLowerCase() in CompoundTokenType) {
+      const compoundService = new CompoundService(collateral.address, collateral.symbol, provider, account)
+      setCompoundService(compoundService)
+      fetchCompoundService()
+    }
+    // eslint-disable-next-line
+  }, [])
```

Comment:
Do we need the `useEffect` here? Doesn't `useMemo` run on mount?

Reply:
yeah useMemo does indeed run on mount, however I tested without it and the app doesn't get the initial values of compoundExchangeRate.
I assume it is because it uses a memoized value of collateral and since it called in the entry level component to display the top section value market_status_open the memoized value of collateral is still null,
That component (market_status_open and market_status_closed) uses the on mount effect to get the values.
Everywhere else the value is fetched using the momized value returned by useMemo.